satellite/satellitedb/overlaycache_test.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package satellitedb
+
+import (
+	"fmt"
+	"math"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zaptest"
+
+	"storj.io/storj/internal/testcontext"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	dbx "storj.io/storj/satellite/satellitedb/dbx"
+)
+
+func TestOverlaycache_AllPieceCounts(t *testing.T) {
+	ctx := testcontext.New(t)
+
+	// setup satellite db
+	db, err := NewInMemory(zaptest.NewLogger(t))
+	require.NoError(t, err)
+	require.NotNil(t, db)
+
+	err = db.CreateTables()
+	require.NoError(t, err)
+
+	// get overlay db
+	lockedOverlay, ok := db.OverlayCache().(*lockedOverlayCache)
+	require.True(t, ok)
+	require.NotNil(t, lockedOverlay)
+
+	overlay, ok := lockedOverlay.db.(*overlaycache)
+	require.True(t, ok)
+	require.NotNil(t, lockedOverlay)
+
+	// create test nodes in overlay db
+	testNodes := createTestNodes(10, t, ctx, overlay.db)
+
+	// update piece count fields on test nodes; set exponentially
+	var (
+		sqlQuery string
+		args     []interface{}
+	)
+	expectedPieceCounts := make(map[storj.NodeID]int)
+	updateSQL := "UPDATE nodes SET ( piece_count ) = ( %d ) WHERE id == ?;"
+	for i, node := range testNodes {
+		pieceCount := int(math.Pow10(i))
+		nodeID, err := storj.NodeIDFromBytes(node.Id)
+		require.NoError(t, err)
+		require.NotEqual(t, storj.NodeID{}, nodeID)
+
+		expectedPieceCounts[nodeID] = pieceCount
+		sqlQuery += fmt.Sprintf(updateSQL+"\n", pieceCount)
+		args = append(args, node.Id)
+	}
+	_, err = overlay.db.DB.ExecContext(ctx, overlay.db.Rebind(sqlQuery), args...)
+	require.NoError(t, err)
+
+	// expected and actual piece count maps should match
+	actualPieceCounts, err := overlay.AllPieceCounts(ctx)
+	require.NoError(t, err)
+	require.Equal(t, expectedPieceCounts, actualPieceCounts)
+}
+
+func TestOverlaycache_UpdatePieceCounts(t *testing.T) {
+	ctx := testcontext.New(t)
+
+	// setup satellite db
+	db, err := NewInMemory(zaptest.NewLogger(t))
+	require.NoError(t, err)
+	require.NotNil(t, db)
+
+	err = db.CreateTables()
+	require.NoError(t, err)
+
+	// get overlay db
+	lockedOverlay, ok := db.OverlayCache().(*lockedOverlayCache)
+	require.True(t, ok)
+	require.NotNil(t, lockedOverlay)
+
+	overlay, ok := lockedOverlay.db.(*overlaycache)
+	require.True(t, ok)
+	require.NotNil(t, lockedOverlay)
+
+	// create test nodes in overlay db
+	testNodes := createTestNodes(10, t, ctx, overlay.db)
+
+	// set expected piece counts
+	expectedPieceCounts := make(map[storj.NodeID]int)
+	for i, node := range testNodes {
+		pieceCount := int(math.Pow10(i))
+
+		var nodeID storj.NodeID
+		copy(nodeID[:], node.Id)
+		expectedPieceCounts[nodeID] = pieceCount
+	}
+
+	// update piece count fields on test nodes; set exponentially
+	err = overlay.UpdatePieceCounts(ctx, expectedPieceCounts)
+	require.NoError(t, err)
+
+	// expected and actual piece count maps should match
+	actualPieceCounts, err := overlay.AllPieceCounts(ctx)
```

Comment:
should we directly query for each node rather than using `AllPieceCounts`?

Reply:
Yes, I had the same thought in bed last night 😛